Flycheck for C++

I have a file "framework.cpp"
with single line`#include "framework.h"`
`C-c ! l` shows next:

```
     1     warning         In include ./framework.h (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
```
Switching to file `framework.h`
```
1   9 warning         #pragma once in main file (c/c++-clang)
     3     error           In include ./render/renderer.h (c/c++-clang)
     3     error           In file included from (c/c++-clang)
```
Switching to file `render/renderer.h`
```
1   9 warning         #pragma once in main file (c/c++-clang)
     3     warning         In include ./device/device.h (c/c++-clang)
     4     error           In file included from (c/c++-clang)
     4     error           In file included from (c/c++-clang)
     4     warning         In include ./draw_stages/draw_stages.h (c/c++-clang)
     4     error           In file included from (c/c++-clang)
     4     error           In file included from (c/c++-clang)
     4     error           In file included from (c/c++-clang)
     4     error           In file included from (c/c++-clang)
     4     error           In file included from (c/c++-clang)
     4     error           In file included from (c/c++-clang)
     4     error           In file included from (c/c++-clang)
     5     error           In file included from (c/c++-clang)
     5     error           In file included from (c/c++-clang)
     5     error           In file included from (c/c++-clang)
     5     error           In file included from (c/c++-clang)
     5     error           In file included from (c/c++-clang)
     5     error           In file included from (c/c++-clang)
     5     error           In file included from (c/c++-clang)
     5     error           In file included from (c/c++-clang)
     5     error           In file included from (c/c++-clang)
     5     error           In file included from (c/c++-clang)
     5     error           In file included from (c/c++-clang)
     5     error           In file included from (c/c++-clang)
     5     error           In file included from (c/c++-clang)
     5     error           In file included from (c/c++-clang)
     5     error           In file included from (c/c++-clang)
     5     error           In file included from (c/c++-clang)
     5     error           In file included from (c/c++-clang)
     5     error           In file included from (c/c++-clang)
     5     error           In file included from (c/c++-clang)
     5     error           In file included from (c/c++-clang)
     5     error           In file included from (c/c++-clang)
     5     error           In file included from (c/c++-clang)
     5     error           In file included from (c/c++-clang)
```
And etc.. etc... etc..
And finally in the some file
```
1   9 warning         #pragma once in main file (c/c++-clang)
     4  10 error           'cstring' file not found (c/c++-clang)
```
```
#pragma once

#include <vulkan/vulkan.h>
#include <cstring>

//more code
```
What I expected: no 100 deep file routines, no useless spam like one hundred "in file included from" without "PATH/FILE_THAT_INCLUDED", and finally `#include <cstring>` mustn't be an error.

`M-x flycheck-verify-setup`
```
Syntax checkers for buffer misc.h in c-mode:

  c/c++-clang
    - may enable: yes
    - executable: Found at /usr/lib/llvm/7/bin/clang

  c/c++-gcc
    - may enable: yes
    - executable: Found at /usr/bin/gcc

  c/c++-cppcheck
    - may enable: Automatically disabled!
    - executable: Not found

Flycheck Mode is enabled.  Use C-u C-c ! x to enable disabled
checkers.

--------------------

Flycheck version: 31
Emacs version:    26.1
System:           x86_64-pc-linux-gnu
Window system:    x
```
System `gentoo`

Also this config didn't work when I call `flycheck-compile c/c++-clang`, please tell me why.
```
;;; Directory Local Variables
;;; For more information see (info "(emacs) Directory Variables")

((c++mode .
	  (
	      (flycheck-clang-language-standard . c++17)
	      (flycheck-clacn-no-exceptions . t)
	      (flycheck-clang-no-rtti . t)
	      (flycheck-clang-warnings .  "everything no-pragma-once-outside-header no-missing-braces no-c++98-compat-pedantic no-nested-anon-types no-zero-length-array no-c++98-c++11-compat-binary-literal no-comma no-c++98-compat no-gnu-anonymous-struct no-c99-extensions no-vla-extension no-c++17-extensions no-switch-enum no-vla")
	      (flycheck-clang-args . nil)
	      )
))
```

> > ((c++mode
> 
> should be `c++-mode`.
> 
Okay, thanks, variables works now. But how to set `c++-mode` by default in `.dir-locals.el`? Maybe better to install project manager to emacs?
> > I am correctly understand, that flycheck can't show output about different files?
> 
> Not sure what you mean. You want errors for different files to appear in `C-c ! l`?
When I am working with templates, template instantiation can cause errors from template files, that don't shown in template file, because only instantiation usage can provide errors, for example.

Okay, i describe this. I have vector library, that defines template vector with overloaded operator +=, looks like this
```

template<unsigned int size, typename type>
struct vector : public vector_base<size, type>
{
    template<typename other_t> constexpr inline
    vector& operator += (const vector<size, other_t>& other);

    type _[size];
}
template<unsigned int size, typename type> template<typename other_t> constexpr inline 
vector<size, type>& vector<size, type>::operator += (const vector<size, other_t>& other)
{
     for (unsigned int i = 0; i < size; ++i)
         _[i] += other[i];
     return *this;
}
```
And duo this, `type` member must have += operator.
Now, I instantiate in some other file this template:
```
struct dummy {};
math::vector<3, dummy> vec1, vec2;
vec1 += vec2;
```
struct dummy doesn't have overloaded operator +=, so
`C-c ! l` shows me 
```
In include ./system.h

In include ./system.h

In include ./system.h

In include ./system.h

In include ./system.h

In include ./system.h

In include ./system.h

In include ./system.h

In include ./system.h
```
In this error QtCreator, as an example of imitation, shows this information

```
base.hpp:68:18: error: no viable overloaded '+='
--
system.cpp:1:10: note: in file included from /home/larry/game/game/system/system.cpp:1:
system.h:3:10: note: in file included from /home/larry/game/game/system/system.h:3:
renderer.h:4:10: note: in file included from /home/larry/game/game/system/../render/renderer.h:4:
draw_stages.h:12:10: note: in file included from /home/larry/game/game/system/../render/draw_data/../draw_stages/draw_stages.h:12:
offscreen.h:8:10: note: in file included from /home/larry/game/game/system/../render/draw_stages/offscreen/offscreen.h:8:
descriptors.h:3:10: note: in file included from /home/larry/game/game/system/../render/draw_stages/offscreen/descriptors.h:3:
math.h:6:10: note: in file included from /home/larry/game/game/system/../math/math.h:6:
...
noise_test.hpp:215:53: note: in instantiation of function template specialization 'seeds::generate_texture_icosagedron_quad<seeds::noise_test_type::GREEN_PLANET>' requested here
system.cpp:163:17: note: in instantiation of function template specialization 'seeds::noise_test_init<seeds::noise_test_type::GREEN_PLANET>' requested here
system.cpp:11:26: note: in instantiation of function template specialization 'System::Object::init<seeds::noise_test_type::GREEN_PLANET>' requested here

```
I can see full trace and click to any line for "go to statement".

> https://github.com/JackalopeGit/b4343j423

Thanks, that's really helpful.  So here's what I'm seeing when running `clang main.cpp`:

```

In file included from main.cpp:1:
In file included from ./template.h:3:
In file included from ./vector.h:4:
./vector_imp.hpp:8:28: error: type 'const vector<3, Dummy>' does not provide a subscript operator
        member_array[i] += other[i];
                           ^~~~~ ~
./template.h:19:39: note: in instantiation of member function 'vector<3, Dummy>::operator+=' requested here
        case FunctionEnum::Add:   vec1+=vec2;
                                      ^
main.cpp:5:5: note: in instantiation of function template specialization 'foo<FunctionEnum::Dummy>' requested here
    foo<FunctionEnum::Dummy>();
    ^
In file included from main.cpp:1:
In file included from ./template.h:3:
In file included from ./vector.h:4:
./vector_imp.hpp:8:28: error: type 'const vector<3, Dummy>' does not provide a subscript operator
        member_array[i] += other[i];
                           ^~~~~ ~
./template.h:19:39: note: in instantiation of member function 'vector<3, Dummy>::operator+=' requested here
        case FunctionEnum::Add:   vec1+=vec2;
                                      ^
main.cpp:6:5: note: in instantiation of function template specialization 'foo<FunctionEnum::Add>' requested here
    foo<FunctionEnum::Add>();
    ^
2 errors generated.
```

Basically, there are **two actual errors** reported by clang, and the other locations are just notes, helpfully pointing to the error *causes*.

Now, here is what the Flycheck error list reports:

![flycheck](https://user-images.githubusercontent.com/340129/47443449-9c3b0980-d7b4-11e8-93cf-555c2b468e9f.png)

Flycheck finds **8 errors** and 2 info diagnostics.  What we would like to see instead is **2 errors** and **4 info**, each info corresponding to the causes of the 2 errors.  In particular, the top 6 errors reported by Flycheck are basically noise (and this noise can get quite loud, as your first post shows).

I've sent #1513 to fix it.

--- a/doc/user/error-interaction.rst
+++ b/doc/user/error-interaction.rst
@@ -138,6 +138,26 @@ and IDs of all errors under the pointer, but the contents are customisable:
    The function is given a list of Flycheck errors to display as sole argument
    and shall return a single string to use as the contents of the tooltip.
 
+Errors from other files
+=======================
+
+Some checkers may return errors from files other than the current buffer (e.g.,
+`gcc` may complain about errors in included files).  These errors appear in the
+error list, and are also added on the first line of the current buffer.  You can
+jump to the incriminating files with `flycheck-previous-error`.
+
+By default, warnings and info messages from other files are ignored, but you can
+customize the minimum level:
+
+.. defcustom:: flycheck-relevant-error-other-file-minimum-level
+
+   The minimum level errors from other files to consider for inclusion in the
+   current buffer.
+
+   If set to an error level, only display errors from other files whose error
+   level is at least as severe as this one.  If ``nil``, display all errors from
+   other files.
+
 Explain errors
 ==============
 

--- a/flycheck.el
+++ b/flycheck.el
@@ -693,6 +693,17 @@ using \\[flycheck-error-list-reset-filter]."
   :safe #'flycheck-error-level-p
   :package-version '(flycheck . "0.24"))
 
+(defcustom flycheck-relevant-error-other-file-minimum-level 'error
+  "The minimum level of errors from other files to display in this buffer.
+
+If set to an error level, only display errors from other files
+whose error level is at least as severe as this one.  If nil,
+display all errors from other files."
+  :group 'flycheck
+  :type 'flycheck-minimum-level
+  :safe #'flycheck-error-level-p
+  :package-version '(flycheck . "32"))
+
 (defcustom flycheck-completing-read-function #'completing-read
   "Function to read from minibuffer with completion.
 
@@ -3368,7 +3379,9 @@ Return ERRORS, modified in-place."
     (and file-name
          (or (null buffer-file-name)
              (not (flycheck-same-files-p buffer-file-name file-name)))
-         (eq 'error (flycheck-error-level err)))))
+         (<= (flycheck-error-level-severity
+              flycheck-relevant-error-other-file-minimum-level)
+             (flycheck-error-level-severity (flycheck-error-level err))))))
 
 (defun flycheck-relevant-error-p (err)
   "Determine whether ERR is relevant for the current buffer.

--- a/flycheck.el
+++ b/flycheck.el
@@ -6764,20 +6764,13 @@ Requires GCC 4.4 or newer.  See URL `https://gcc.gnu.org/'."
             "-")
   :standard-input t
   :error-patterns
-  ((error line-start
-          (message "In file included from") " " (or "<stdin>" (file-name))
-          ":" line ":" column ":" line-end)
-   (info line-start (or "<stdin>" (file-name)) ":" line ":" column
+  ((info line-start (or "<stdin>" (file-name)) ":" line ":" column
          ": note: " (message) line-end)
    (warning line-start (or "<stdin>" (file-name)) ":" line ":" column
             ": warning: " (message (one-or-more (not (any "\n["))))
             (optional "[" (id (one-or-more not-newline)) "]") line-end)
    (error line-start (or "<stdin>" (file-name)) ":" line ":" column
           ": " (or "fatal error" "error") ": " (message) line-end))
-  :error-filter
-  (lambda (errors)
-    (flycheck-fold-include-levels (flycheck-sanitize-errors errors)
-                                  "In file included from"))
   :modes (c-mode c++-mode)
   :next-checkers ((warning . c/c++-cppcheck)))
 

--- a/flycheck.el
+++ b/flycheck.el
@@ -6607,10 +6607,7 @@ See URL `http://clang.llvm.org/'."
             "-")
   :standard-input t
   :error-patterns
-  ((error line-start
-          (message "In file included from") " " (or "<stdin>" (file-name))
-          ":" line ":" line-end)
-   (info line-start (or "<stdin>" (file-name)) ":" line ":" column
+  ((info line-start (or "<stdin>" (file-name)) ":" line ":" column
          ": note: " (optional (message)) line-end)
    (warning line-start (or "<stdin>" (file-name)) ":" line ":" column
             ": warning: " (optional (message)) line-end)
@@ -6625,7 +6622,7 @@ See URL `http://clang.llvm.org/'."
         ;; them past our error filtering
         (setf (flycheck-error-message err)
               (or (flycheck-error-message err) "no message")))
-      (flycheck-fold-include-levels errors "In file included from")))
+      errors))
   :modes (c-mode c++-mode)
   :next-checkers ((warning . c/c++-cppcheck)))
 

--- a/test/flycheck-test.el
+++ b/test/flycheck-test.el
@@ -2936,12 +2936,21 @@ evaluating BODY."
          :id "-Wsign-compare" :checker c/c++-gcc)
      '(8 7 warning "#warning" :id "-Wcpp" :checker c/c++-gcc))))
 
-(flycheck-ert-def-checker-test c/c++-gcc (c c++) included-file-error
+(flycheck-ert-def-checker-test c/c++-gcc (c c++) included-file-warning
   (let ((flycheck-gcc-include-path '("./include"))
-        (flycheck-disabled-checkers '(c/c++-clang)))
+        (flycheck-disabled-checkers '(c/c++-clang))
+        (flycheck-relevant-error-other-file-minimum-level 'warning))
     (flycheck-ert-should-syntax-check
      "language/c_c++/in-included-file.cpp" 'c++-mode
-     `(3 nil warning "In include warning.c" :checker c/c++-gcc))))
+     `(5 10 warning "unused variable 'unused'"
+         :filename ,(flycheck-ert-resource-filename "language/c_c++/warning.c")
+         :id "-Wunused-variable" :checker c/c++-gcc)
+     `(7 15 warning "comparison between signed and unsigned integer expressions"
+         :filename ,(flycheck-ert-resource-filename "language/c_c++/warning.c")
+         :id "-Wsign-compare" :checker c/c++-gcc)
+     `(8 7 warning "#warning"
+         :filename ,(flycheck-ert-resource-filename "language/c_c++/warning.c")
+         :id "-Wcpp" :checker c/c++-gcc))))
 
 (flycheck-ert-def-checker-test c/c++-cppcheck (c c++) nil
   :tags '(cppcheck-xml)

--- a/test/flycheck-test.el
+++ b/test/flycheck-test.el
@@ -2896,12 +2896,21 @@ evaluating BODY."
          :checker c/c++-clang)
      '(8 7 warning "no message" :checker c/c++-clang))))
 
-(flycheck-ert-def-checker-test c/c++-clang (c c++) included-file-error
+(flycheck-ert-def-checker-test c/c++-clang (c c++) included-file-warning
   (let ((flycheck-clang-include-path '("./include"))
-        (flycheck-disabled-checkers '(c/c++-gcc)))
+        (flycheck-disabled-checkers '(c/c++-gcc))
+        (flycheck-relevant-error-other-file-minimum-level 'warning))
     (flycheck-ert-should-syntax-check
      "language/c_c++/in-included-file.cpp" 'c++-mode
-     `(3 nil warning "In include ./warning.c" :checker c/c++-clang))))
+     `(5 10 warning "unused variable 'unused'"
+         :filename ,(flycheck-ert-resource-filename "language/c_c++/warning.c")
+         :checker c/c++-clang)
+     `(7 15 warning "comparison of integers of different signs: 'int' and 'unsigned int'"
+         :filename ,(flycheck-ert-resource-filename "language/c_c++/warning.c")
+         :checker c/c++-clang)
+     `(8 7 warning "no message"
+         :filename ,(flycheck-ert-resource-filename "language/c_c++/warning.c")
+         :checker c/c++-clang))))
 
 (flycheck-ert-def-checker-test c/c++-gcc (c c++) error
   (let ((flycheck-disabled-checkers '(c/c++-clang)))

